core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala

```diff
@@ -137,6 +139,17 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
     )
   }
 
+  /** If Tasks is too large, remove and garbage collect old tasks */
+  private def trimTasksIfNecessary(taskData: HashMap[Long, TaskUIData]) = synchronized {
+    if (taskData.size > retainedTasks) {
+      val toRemove = (taskData.size - retainedTasks)
+      val oldIds = taskData.map(_._2.taskInfo.taskId).toList.sorted.take(toRemove)
```

Comment:
This code is from the original pr, I didn't know _._1 is always equal to _._2.taskInfo.taskId so I'll change that, as for the loop, it does slow things down, I noticed it in my testing, I'll look at a better way to do this.